### PR TITLE
[New arch] Use moshi to parse sharees

### PR DESCRIPTION
**Pull request by theScrabi**
related Issue: https://github.com/owncloud/android/issues/2704
related lib pr: https://github.com/owncloud/android-library/pull/352

## QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/Android/Release_2.16/2994-sharees_moshi.md

### CODE REVIEW

File: owncloudData/src/main/java/com/owncloud/android/data/sharing/sharees/datasources/mapper/RemoteShareeMapper.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.owncloud.android.data.sharing.sharees.datasources.mapper
```
Comment (abelgardep) [inline owncloudData/src/main/java/com/owncloud/android/data/sharing/sharees/datasources/mapper/RemoteShareeMapper.kt:1]: Add license header and a newline at end of file 👍

File: owncloudData/src/main/java/com/owncloud/android/data/sharing/sharees/datasources/mapper/RemoteShareeMapper.kt [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.owncloud.android.domain.mappers.RemoteMapper
+import com.owncloud.android.domain.sharing.sharees.model.OCSharee
+import com.owncloud.android.domain.sharing.shares.model.ShareType
+import com.owncloud.android.lib.resources.shares.responses.ShareeItem
+import com.owncloud.android.lib.resources.shares.responses.ShareeOcsResponse
+
+class RemoteShareeMapper : RemoteMapper<List<OCSharee>, ShareeOcsResponse> {
+    private fun mapShareeItemToOCShare(item: ShareeItem, isExactMatch: Boolean) =
```
Comment (abelgardep) [inline owncloudData/src/main/java/com/owncloud/android/data/sharing/sharees/datasources/mapper/RemoteShareeMapper.kt:10]:
```suggestion
    private fun mapShareeItemToOCSharee(item: ShareeItem, isExactMatch: Boolean) =
```

File: owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt [hunk 3]
```diff
@@ -20,40 +20,37 @@
...
 import org.junit.Assert.assertEquals
+import org.junit.Assert.assertFalse
 import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertTrue
 import org.junit.Before
-import org.junit.Ignore
 import org.junit.Test
 
 // Needs to be instrumented since JSONObject is tied to the Android platform
 class OCRemoteShareesDataSourceTest {
```
Comment (abelgardep) [inline owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt:42]: This test is inside shares package, i think it should be included in sharees package, what do you think?

File: owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt [hunk 4]
```diff
@@ -20,40 +20,37 @@
...
-import org.json.JSONObject
 import org.junit.Assert.assertEquals
+import org.junit.Assert.assertFalse
 import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertTrue
 import org.junit.Before
-import org.junit.Ignore
 import org.junit.Test
 
 // Needs to be instrumented since JSONObject is tied to the Android platform
```
Comment (abelgardep) [inline owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt:41]: This comment is not needed anymore from my point of view.

File: owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt [hunk 5]
```diff
@@ -20,40 +20,37 @@
...
 class OCRemoteShareesDataSourceTest {
     private lateinit var ocRemoteShareesDataSource: OCRemoteShareeDataSource
     private val ocShareeService: OCShareeService = mockk()
+    private lateinit var sharees: List<OCSharee>;
```
Comment (abelgardep) [inline owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt:45]: Kotlin 🎉   No need to add `;`

File: owncloudDomain/src/main/java/com/owncloud/android/domain/sharing/sharees/GetShareesAsyncUseCase.kt [hunk 6]
```diff
@@ -20,12 +20,13 @@
 package com.owncloud.android.domain.sharing.sharees
 
 import com.owncloud.android.domain.BaseUseCaseWithResult
+import com.owncloud.android.domain.sharing.sharees.model.OCSharee
 import org.json.JSONObject
```
Comment (abelgardep) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/sharing/sharees/GetShareesAsyncUseCase.kt:24]: JSONObject is not needed anymore, is it?

File: owncloudDomain/src/main/java/com/owncloud/android/domain/sharing/sharees/model/OCSharee.kt [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+import com.owncloud.android.domain.sharing.shares.model.ShareType
+
+data class OCSharee(
+    val label: String,
+    val shareType: ShareType,
+    val shareWith: String,
+    val additionalInfo: String,
+    val isExactMatch: Boolean
+)
```
Comment (abelgardep) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/sharing/sharees/model/OCSharee.kt:11]: License header and newline at end of file

File: owncloudDomain/src/main/java/com/owncloud/android/domain/sharing/shares/model/OCShare.kt [hunk 8]
```diff
@@ -71,17 +71,7 @@ enum class ShareType constructor(val value: Int) {
...
-                1 -> GROUP
-                3 -> PUBLIC_LINK
-                4 -> EMAIL
-                5 -> CONTACT
-                6 -> FEDERATED
-                else -> null
-            }
-        }
+        private val values = values();
+        fun fromValue(value: Int) = values.firstOrNull { it.value == value }
```
Comment (abelgardep) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/sharing/shares/model/OCShare.kt:75]:
```suggestion
        fun fromValue(value: Int) = values().firstOrNull { it.value == value }
```

File: owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt [hunk 9]
```diff
@@ -62,54 +59,83 @@ class OCRemoteShareesDataSourceTest {
         } returns getRemoteShareesOperationResult
 
         // Get sharees from remote datasource
-        val sharees = ocRemoteShareesDataSource.getSharees(
+        sharees = ocRemoteShareesDataSource.getSharees(
             "user",
             1,
             30
```
Comment (abelgardep) [inline owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt:65]: I'm not sure at all about this test. `sharees` will be always the same mocked list, what happens when no sharees are retrieved? Is it an empty list or just null? Maybe some edge cases scenarios are missing, what do you think?

**Review by abelgardep — CHANGES_REQUESTED** [review]
Looks good but some changes requested @theScrabi

File: owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt (hunk 9 above)
Comment (theScrabi) [reply owncloudData/src/test/java/com/owncloud/android/data/shares/datasources/OCRemoteShareesDataSourceTest.kt:65]: yes

**Review by abelgardep — APPROVED** [review]
Approved, but rebase it with master before moving it to QA please.

**Comment by jesmrec** [comment]
Approved an my side
